Implement new Import CVR Modal in Election Manager

carolinemodic opened the pull request:
This PR implements a new import CVR Modal in election manager, automatically searching for the automatically exported files from https://github.com/votingworks/vxsuite/pull/109

There are a few things this PR does NOT do, including any changes to the Tally results table or the errors/warnings that get shown there, no errors or warnings are shown in this dialog as soon as you import the dialog closes and the results show up in the tally table. Possible future improvements to that piece are listed in the zube ticket: https://zube.io/votingworks/vxsuite/c/2848

When importing and no files are automagically found, and none have been previously imported:
![Screen Shot 2020-12-15 at 1 11 42 PM](https://user-images.githubusercontent.com/14897017/102285955-1e826c80-3eec-11eb-89e1-5492e275bd07.png)

When importing and there are already live mode CVRs imported that are also still found on the USB: (regardless if there are also test mode files on the usb)
![Screen Shot 2020-12-15 at 1 15 51 PM](https://user-images.githubusercontent.com/14897017/102285623-84babf80-3eeb-11eb-95d5-67e9122543b1.png)

When importing when there is nothing already imported but only live mode CVRs are found on the usb:
![Screen Shot 2020-12-15 at 1 14 57 PM](https://user-images.githubusercontent.com/14897017/102285629-871d1980-3eeb-11eb-88d6-c4c32fc2726e.png)

When importing and all files on the usb were already imported: 
![Screen Shot 2020-12-15 at 1 13 50 PM](https://user-images.githubusercontent.com/14897017/102285631-87b5b000-3eeb-11eb-88e0-b45fe9faf1d2.png)

When test mode CVRs were already imported (and are still on the USB) but there are new files on the USB (regardless if there are also live mode files on the usb)
![Screen Shot 2020-12-15 at 1 13 40 PM](https://user-images.githubusercontent.com/14897017/102285633-884e4680-3eeb-11eb-9c47-139c819d7e28.png)

When importing and there are only test mode CVRs found on the USB but no files were previously imported: 
![Screen Shot 2020-12-15 at 1 13 16 PM](https://user-images.githubusercontent.com/14897017/102285636-884e4680-3eeb-11eb-9b60-ab1f712aabba.png)

When importing with nothing already imported and both live and test mode files on the USB: 
![Screen Shot 2020-12-15 at 1 12 36 PM](https://user-images.githubusercontent.com/14897017/102285637-88e6dd00-3eeb-11eb-81e1-069756210669.png)
![Screen Shot 2020-12-15 at 1 12 29 PM](https://user-images.githubusercontent.com/14897017/102285883-027ecb00-3eec-11eb-869a-88836f8b7a74.png)

And when there is no usb present: 
![Screen Shot 2020-12-15 at 3 43 44 PM](https://user-images.githubusercontent.com/14897017/102286056-54bfec00-3eec-11eb-94fb-16cc01df8142.png)

carolinemodic commented:
Gif of some of the screens in action: 

![importcvrs2](https://user-images.githubusercontent.com/14897017/102287434-4cb57b80-3eef-11eb-8700-a464954decd8.gif)

eventualbuddha commented:
Thanks for all the screenshots and the gif. Looking forward to digging into this.

eventualbuddha commented on apps/election-manager/src/utils/filenames.ts:
```diff
@@ -7,26 +8,86 @@ const SUBSECTION_SEPARATOR = '_'
+  )}`
+}
+
+export function parseCVRFileInfoFromFilename(
+  filename: string
+): CVRFileData | undefined {
+  const segments = filename.split(SECTION_SEPARATOR)
+  const isTestModeResults = segments.length === 4 && segments[0] === 'TEST'
+
+  const postTestPrefixSegments = isTestModeResults
+    ? segments.slice(1)
+    : segments
+  if (postTestPrefixSegments.length !== 3) {
+    return
+  }
+
+  const machineSegments = postTestPrefixSegments[0].split(SUBSECTION_SEPARATOR)
+  if (machineSegments.length !== 2 || machineSegments[0] !== 'machine') {
+    return
+  }
+  const machineId = machineSegments[1]
+
+  const ballotSegments = postTestPrefixSegments[1].split(SUBSECTION_SEPARATOR)
+  if (ballotSegments.length !== 2 || ballotSegments[1] !== 'ballots') {
+    return
+  }
+  const numberOfBallots = Number(ballotSegments[0])
+
+  const parsedTime = moment(postTestPrefixSegments[2], TIME_FORMAT_STRING)
+  return {
+    machineId,
+    numberOfBallots,
+    isTestModeResults,
+    timestamp: parsedTime.toDate(),
+  }
+}
```
Another approach, not necessarily better, but I felt compelled to try something:

```suggestion
const CVR_FILE_PATTERN = /^(TEST__)?machine_([^_]+?)__(\d+?)_ballots__(\d{4})-(\d{2})-(\d{2})(?:_(\d{2})-(\d{2})-(\d{2}))?\.jsonl$/

export function parseCVRFileInfoFromFilename(
  filename: string
): CVRFileData | undefined {
  const match = filename.match(CVR_FILE_PATTERN)

  if (!match) {
    return
  }

  const [
    ,
    test,
    machineId,
    numberOfBallots,
    year,
    month,
    day,
    hour,
    minute,
    second,
  ] = match
  return {
    isTestModeResults: !!test,
    machineId,
    numberOfBallots: Number(numberOfBallots),
    timestamp:
      hour === undefined
        ? new Date(Number(year), Number(month) - 1, Number(day))
        : new Date(
            Number(year),
            Number(month) - 1,
            Number(day),
            Number(hour),
            Number(minute),
            Number(second)
          ),
  }
}
```

eventualbuddha commented on apps/election-manager/src/utils/CastVoteRecordFiles.ts:
```diff
@@ -278,6 +336,31 @@ export default class CastVoteRecordFiles {
+  }
+
+  public filenameAlreadyImported(filename: string): boolean {
+    for (const f of this.files) {
+      if (f.name === filename) {
```
```suggestion
    for (const file of this.files) {
      if (file.name === filename) {
```

Or, if you don't want to name the variable:

```suggestion
    for (const { name } of this.files) {
      if (name === filename) {
```

eventualbuddha commented on apps/election-manager/src/utils/filenames.test.ts:
```diff
@@ -64,3 +67,89 @@ test('generates ballot export package name with zero padded time pieces', () =>
+  test('parses a basic name not in test mode properly', () => {
+    const name = 'machine_5__1_ballots__2020-12-08_10-42-02.jsonl'
+    const results = parseCVRFileInfoFromFilename(name)
+    expect(results)
+    expect(results?.machineId).toBe('5')
+    expect(results?.numberOfBallots).toBe(1)
+    expect(results?.isTestModeResults).toBe(false)
+    expect(results?.timestamp).toStrictEqual(new Date(2020, 11, 8, 10, 42, 2))
```
I usually find it's better to do these assertions in one go so you get more context around a failure.
```suggestion
    expect(results).toEqual({
      isTestModeResults: false,
      machineId: '5',
      numberOfBallots: 1,
      timestamp: new Date(2020, 11, 8, 10, 42, 2),
    })
```

eventualbuddha reviewed (commented):
First thoughts. Exciting!

eventualbuddha commented on apps/election-manager/src/components/ImportCVRFilesModal.tsx:
```diff
@@ -0,0 +1,365 @@
+    const headerModeText =
+      fileMode === 'test' ? 'Test Mode' : fileMode === 'live' ? 'Live Mode' : ''
+
+    let instructionalText = null
```
```suggestion
    let instructionalText: string
```

eventualbuddha reviewed (commented):
More comments on minor stuff.

Bigger picture, I think this is too verbose. Mostly, nobody's going to read or understand this stuff. Can we make the test/live thing less of an issue? I still think having an explicit "start tabulating election" step could include a toggle there.

carolinemodic commented:
Updated the design to instead focus around 1 at a time file import and updated the screenshots in the original PR request. New video is shown below. I removed the sentence in the modal about why live mode CVR files aren't being shown once test mode files have been imported (and vice versa) as that is needlessly confusing but could add that back (or a similar clearer sentence). 

Also when you are "locked" into a file mode added this sentence on the main tally page
Test Mode:
![Screen Shot 2021-01-05 at 10 45 24 AM](https://user-images.githubusercontent.com/14897017/103686117-2bbad600-4f43-11eb-870e-a70cd044008c.png)

Live Mode:
![Screen Shot 2021-01-05 at 10 45 39 AM](https://user-images.githubusercontent.com/14897017/103686129-31182080-4f43-11eb-8ed2-73e7343dcda6.png)

New Video:

https://user-images.githubusercontent.com/14897017/103686647-febaf300-4f43-11eb-91bb-039fdf346180.mov

cc @beausmith @eventualbuddha

beausmith reviewed (approved):
This is great @carolinemodic!! Thanks for the video walkthrough and thanks for persisting through all the feedback whiplash. This UI is certainly going to delight election admins. 😸

beausmith commented:
@carolinemodic - Just wanted to also note that I agree with @eventualbuddha's code optimization comments. 👍
